teachertool/src/components/SplitPane.tsx

```diff
@@ -5,20 +5,115 @@ import { classList } from "react-common/components/util";
+    const [size, setSize] = React.useState(startingSize ?? defaultSize);
+    const [isResizing, setIsResizing] = React.useState(false);
+    const containerRef = React.useRef<HTMLDivElement>(null);
+
+    React.useEffect(() => {
+        if (!isResizing) {
+            onResizeEnd?.(size);
+        }
+    }, [isResizing]);
+
+    function handleResizeMouse(event: MouseEvent) {
+        handleResize(event.clientX, event.clientY);
+    }
+
+    function handleResizeTouch(event: TouchEvent) {
+        if (event.touches.length >= 1) {
+            handleResize(event.touches[0].clientX, event.touches[0].clientY);
+        }
+    }
+
+    function handleResize(clientX: number, clientY: number) {
+        const containerRect = containerRef.current?.getBoundingClientRect();
+        if (containerRect) {
+            const newSize =
+                split === "vertical"
+                    ? `${((clientX - containerRect.left) / containerRect.width) * 100}%`
+                    : `${((clientY - containerRect.top) / containerRect.height) * 100}%`;
+
+            setIsResizing(true); // Do this here rather than inside startResizing to prevent interference with double click detection.
+            setSize(newSize);
+        }
+    }
+
+    function startResizing(event: React.MouseEvent | React.TouchEvent) {
+        event.preventDefault();
+        document.addEventListener("mousemove", handleResizeMouse);
+        document.addEventListener("mouseup", endResizing);
+        document.addEventListener("touchmove", handleResizeTouch);
+        document.addEventListener("touchend", endResizing);
+    }
+
+    function endResizing() {
+        document.removeEventListener("mousemove", handleResizeMouse);
+        document.removeEventListener("mouseup", endResizing);
+        document.removeEventListener("touchmove", handleResizeTouch);
+        document.removeEventListener("touchend", endResizing);
+
+        setIsResizing(false);
+    }
+
+    function setToDefaultSize() {
+        setSize(defaultSize);
+        onResizeEnd?.(defaultSize);
+    }
+
+    const leftStyle: React.CSSProperties = { flexBasis: size };
+    if (split === "vertical") {
+        leftStyle.minWidth = leftMinSize;
+
+        // Setting right's minWidth doesn't work because left is still allowed
+        // to expand beyond it. Instead, set left's maxWidth.
+        leftStyle.maxWidth = `calc(100% - ${rightMinSize})`;
+    } else {
+        leftStyle.minHeight = leftMinSize;
+        leftStyle.maxHeight = `calc(100% - ${rightMinSize})`;
+    }
+
     return (
-        <div className={classList(css[`split-pane-${split}`], className)}>
-            <div className={css[`left-${split}`]}>{left}</div>
-            <div className={css[`splitter-${split}`]}>
-                <div className={css[`splitter-${split}-inner`]} />
+        <div ref={containerRef} className={classList(css["split-pane"], css[`split-${split}`], className)}>
+            <div className={css[`left`]} style={leftStyle}>
```

Comment:
I like keeping style outside of components as much as possible; I'm wondering if a pattern like this would work in this case. https://github.com/microsoft/pxt/blob/16f4fb291a47583c317410125fe40a30ef7e1ed4/webapp/src/toolbox.tsx#L758
Where we just define css variables with the values we need and then leverage these css variables in the scss modules. I did this in this PR a while back https://github.com/microsoft/pxt/pull/9657, but the pattern might not be the best fit for every scenario.

Reply:
Hmm, we could do it that way, but I'm not sure I fully understand the benefit? We're setting css inside the component either way. Do the variables provide any practical value, or is it just unnecessary complexity? If I understand correctly, I think it made more sense in the HC change because those variables are sometimes set through css and sometimes js. HC is kind of an override. That's not quite the case here, where it's basically always set through JS all the time. I could be missing/mis-understanding something, though.